yellow-paper/docs/contracts/main.md

```diff
@@ -0,0 +1,319 @@
+
+This means that all validation is done by the application circuit. The application should:
+- Ensure that the message exists in the outbox (message tree)
+- Ensure that the user knows `secret` that hashes to the `secretHash` of the message
+- Compute a nullifier that includes the `secret` along with the msg hash and the index of the message in the tree
+    - The index is included to ensure that the nullifier is unique for each message
+
+
+## Logical Execution
+Below, we will outline the **LOGICAL** execution of a L2 block and how the contracts interact with the circuits. We will be executing cross-chain communication before and after the block itself. Note that in reality, the L2 inbox does not exists, and its functionality is handled by the kernel and the rollup circuits.
+
+
+```mermaid
+sequenceDiagram
+    autonumber
+    title Logical Interactions of Crosschain Messages
+
+    participant P2 as Portal (L2)
+
+    participant I2 as Inbox (L2)
+    participant O2 as Outbox (L2)
+    participant R2 as Rollup (L2)
+    participant R as Validating Light Node (L1)
+    participant Reg as Registry
+    participant I as Inbox
+    participant O as Outbox
+
+    participant P as Portal
+
+    P->>I: Send msg to L2
+    I->>I: Populate msg values
+    I->>I: Update state (insert)
+
+    loop block in chain
+
+        loop tx in txs
+
+            loop msg in tx.l1ToL2Consume
+                P2->>O2: Consume msg
+                O2->>O2: Validate msg
+                O2->>O2: Update state (nullify)
+            end
+
+            loop msg in tx.l2ToL1Msgs
+                P2->>I2: Add msg
+                I2->>I2: Populate msg values
+                I2->>I2: Update state (insert)
+            end
+        end
+
+        loop msg in L2 inbox 
+            R2->>O2: Consume msg
+            O2->>O2: Update state (delete)
+        end
+
+        loop msg in l1ToL2Msgs 
+            R2->>O2: Insert msg
+            O2->>O2: Update state (insert)
+        end
+
+        R2->>R: Block (Proof + Data)
+
+        R->>R: Verify proof
+        R->>R: Update State 
+
+        R->>Reg: Where is the Inbox?
+        Reg->>R: Here is the address
+
+        R->>I: Consume l1ToL2Msgs from L1
+        I->>I: Update state (delete)
+
+        R->>Reg: Where is the Outbox?
+        Reg->>R: Here is the address
+
+        R->>O: Insert Messages from L2
+        O->>O: Update state (insert)
+
+    end
+
+    P->>O: Consume a msg
+    O->>O: Validate msg
+    O->>O: Update state (delete)
+```
+We will walk briefly through the steps of the diagram above.
+
+1. A portal contracts on L1 want to send a message for L2
+1. The inbox populates the message with `sender` information
+1. The inbox contract inserts the message into its storage
+1. On the L2, as part of a rollup, a transaction tries to consume a message from the L2 outbox.
```

Comment:
(if the deadline hasn't passed). The transaction may sort all the messages based on fees attached to them